Bootstrap enhancements.

kwoodson opened the pull request:
This goal of this PR is to fix and customize the AWS provisioning.
- This PR allows for custom node groups to be created
- This properly creates the ELB for the infra nodes (this was the blocker)
- This introduces a bootstrap configmap role that allows for more flexibility when creating node configs with less dependencies

kwoodson commented:
@smarterclayton, Please review these changes.  These are very important to the AWS provisioning.

kwoodson commented:
/retest

smarterclayton commented on roles/openshift_node_bootstrap_configmap/defaults/main.yml:
```diff
@@ -0,0 +1,27 @@
+openshift_node_bootstrap_confimap_groups:
+- type: master
+  labels:
+  - 'type=master'
```
As a note here, I opened #6011 because we need to be moving in the right direction with labels.  I would expect us to introduce the new "kube role label" pattern while preserving the old labels for some period of time.  But I think we're going to need to take away the choice (in our install process) to have a different pattern.

smarterclayton commented on roles/openshift_node_bootstrap_configmap/defaults/main.yml:
```diff
@@ -0,0 +1,27 @@
+---
+openshift_node_bootstrap_confimap_groups:
```
Typo, configmap.

Also, if you're going to rename `openshift_master_node_configs` then you should remove it from the master role.

kwoodson replied:
The typo was my fault.

Removing from master was definitely a merge issue.

kwoodson replied:
Fixed.

kwoodson commented:
@smarterclayton, let me know when you think this one is ready.  

I'll run through a test end to end.

kwoodson commented:
Latest install was successful.  3 ha masters /2 infra /3 compute
Install time with provisioning: `0:16:23.592`

kwoodson commented:
@smarterclayton, I have changed to the discussed `openshift_node_group` role with its settings.  I also added some logic surrounding the template and the configmap in an attempt to not stomp the current config. 

If it doesn't exist, template it, apply edits, and create the configmap.  
If it does exist, fetch the configmap, apply edits, and idempotently handle the configmap.

Please take a look

kwoodson commented:
```
INSTALLER STATUS ***************************************************************************************************************************
Initialization             : Complete
Health Check               : Complete
etcd Install               : Complete
Master Install             : Complete
Master Additional Install  : Complete
Node Install               : Complete
Hosted Install             : Complete
Metrics Install            : Complete
Logging Install            : Complete

Tuesday 07 November 2017  00:02:35 -0500 (0:00:31.091)       0:27:26.284 ****** 
```
```
NAME                            STATUS                     AGE       VERSION
ip-172-31-48-113.ec2.internal   Ready                      6h        v1.7.6+a08f5eeb62
ip-172-31-51-105.ec2.internal   Ready                      6h        v1.7.6+a08f5eeb62
ip-172-31-57-196.ec2.internal   Ready,SchedulingDisabled   6h        v1.7.6+a08f5eeb62
ip-172-31-58-143.ec2.internal   Ready,SchedulingDisabled   6h        v1.7.6+a08f5eeb62
ip-172-31-58-162.ec2.internal   Ready                      6h        v1.7.6+a08f5eeb62
ip-172-31-58-250.ec2.internal   Ready                      6h        v1.7.6+a08f5eeb62
ip-172-31-59-170.ec2.internal   Ready,SchedulingDisabled   6h        v1.7.6+a08f5eeb62
ip-172-31-63-196.ec2.internal   Ready                      6h        v1.7.6+a08f5eeb62
```

smarterclayton commented on roles/openshift_node_group/defaults/main.yml:
```diff
@@ -0,0 +1,27 @@
+---
+openshift_node_groups:
+- type: master
```
How are you using `type`?  I would prefer `name` - there is no such thing as `type`.

smarterclayton replied:
Node groups aren't special, but they are identified by name.  Targeting pods and workloads to node groups is done by labels.  Multiple node groups can have the same labels.

kwoodson replied:
I was using it like this in main.yml:
```
    openshift_node_group_name: "{{ openshift_node_group_prefix}}{{ node_group.type }}"
```
I can switch to name if that's preferred.

smarterclayton replied:
Yeah, name would be preferred.

kwoodson replied:
Switched.  Can't test because images aren't available from the docker registries.

kwoodson commented:
/test install

kwoodson commented:
@smarterclayton, I changed it to `oreg_url` if it is defined else `openshift.node.registry_url`.

kwoodson commented:
Tested and everything looks good.  images deployed without any other modifications.

smarterclayton commented:
Verified this was working on GCE - one issue was that prometheus is now showing up as installed by default (no idea how) but that doesn't block this release.

/lgtm

kwoodson commented:
logging test failures are unrelated.

/test logging
